### DISCUSSION TITLE
Support for Multiple AEDT Instances

### DISCUSSION THREAD

**Discussion by atlanswer** [ansys/pyaedt#3916] [General]
Hi. #3342 mentioned that some refactorings have been made to support multiple desktop instances in the future. So I'm wondering:

1. Is there a timeline for that to be implemented?
2. Before it arrives, the end users can try to achieve similar functionality by managing references to the desktop objects themselves. At least there's a workaround that is supposed to work for PyAEDT 0.7.

   Users can clear the `_desktop_sessions` dict after creating one `Desktop` instance to get to `__init__` again. However, if `starting_mode == "grpc"`, the DLL function `pyaedt.generic.grpc_plugin.CreateAedtApplication` seems to return a singleton, i.e., making every desktop wrapper object created to reference the latest one.

   Fallback to COM by `settings.use_grpc_api = False` can work. But currently, a lot of instance variables, methods, and properties are referencing the global variables such as the desktop object `oDesktop` stored in `sys.modules["__main__"]`, which will be overwritten by subsequent initializations of new `Desktop` objects.

   I think one solution is to get rid of global references like `self._main = sys.modules["__main__"]`. For instance, setting the `odesktop` instance variable directly to the AEDT COM object `oDesktop`, or whatever wrapper the corresponding initialization function may return. Right now, instance variables and methods can be overridden by users but properties without setters couldn't.

   Is this solution the intended one? Are issues and PRs that replace these global references with local ones acceptable? For example:

   https://github.com/ansys/pyaedt/blob/09db4b3907850d635a0be750abb07bcd44235422/pyaedt/desktop.py#L1047
   https://github.com/ansys/pyaedt/blob/09db4b3907850d635a0be750abb07bcd44235422/pyaedt/desktop.py#L632
   ```diff
   - self.odesktop = self._main.oDesktop
   + self.odesktop = oApp.GetAppDesktop()
   ```

   https://github.com/ansys/pyaedt/blob/2a62fa5101a3df3ff13000f332917667d48017a8/pyaedt/desktop.py#L650
   ```diff
   - self.aedt_version_id = self._main.oDesktop.GetVersion()[0:6]
   + self.aedt_version_id = self.odesktop.GetVersion()[0:6]
   ```

Thanks. Keep up the good work. 👍

**Comment by Samuelopez-ansys** [comment]
@atlanswer Let us review your second bullet.

Regarding the first one, it is not a PyAEDT limitation, it is an AEDT issue, there is not timeline for that, but the development team is working on that.

@maxcapodi78

**Comment by Samuelopez-ansys** [comment]
We found a limitation in Linux to support multiple Desktop instances. But with this Pull request, it will be possible in Widnows:
https://github.com/ansys/pyaedt/pull/4371

**Reply by atlanswer** [reply]
Cool, thanks for the update.
